Sources/NIOPosix/SelectableEventLoop.swift

```diff
@@ -421,46 +427,80 @@ Further information:
         }
     }
 
+    private func run(_ task: UnderlyingTask) {
+        /* for macOS: in case any calls we make to Foundation put objects into an autoreleasepool */
+        withAutoReleasePool {
+            switch task {
+            case .function(let function):
+                function()
+            #if compiler(>=5.9)
+            case .unownedJob(let erasedUnownedJob):
+                if #available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *) {
+                    erasedUnownedJob.unownedJob.runSynchronously(on: self.asUnownedSerialExecutor())
+                } else {
+                    fatalError("Tried to run an UnownedJob without runtime support")
+                }
+            #endif
+            }
+        }
+    }
+
     /// Start processing I/O and tasks for this `SelectableEventLoop`. This method will continue running (and so block) until the `SelectableEventLoop` is closed.
     internal func run() throws {
         self.preconditionInEventLoop()
         defer {
-            var scheduledTasksCopy = ContiguousArray<ScheduledTask>()
             var iterations = 0
+            var drained = false
             repeat { // We may need to do multiple rounds of this because failing tasks may lead to more work.
-                scheduledTasksCopy.removeAll(keepingCapacity: true)
 
-                self._tasksLock.withLock { () -> Void in
+                let (scheduledTasksCopy, immediateTasksCopy) = self._tasksLock.withLock {
                     // In this state we never want the selector to be woken again, so we pretend we're permanently running.
                     self._pendingTaskPop = true
 
+                    // During the shutdown we don't need to be that careful about reallocs
+                    var scheduledTasksCopy = ContiguousArray<ScheduledTask>()
                     // reserve the correct capacity so we don't need to realloc later on.
                     scheduledTasksCopy.reserveCapacity(self._scheduledTasks.count)
                     while let sched = self._scheduledTasks.pop() {
                         scheduledTasksCopy.append(sched)
                     }
+                    var immediateTasksCopy = ContiguousArray<UnderlyingTask>()
```

Comment:
We should avoid allocating new arrays here on each iteration. Prefer to have persistent ones we reuse. This is a defer block, but the loop may still execute a hefty number of times.

Reply:
Done.